slapp/transforms/video_utils.py

```diff
@@ -46,32 +51,130 @@ def downsample_h5_video(
+            for path in video_paths:
+                fp.write(f"file '{path}'\n")
 
+        # See: https://trac.ffmpeg.org/wiki/Concatenate
+        ffmpeg_concat_cmd = [mpg.get_ffmpeg_exe(), '-y', '-f', 'concat',
+                             '-safe', '0', '-i', concat_list_file.name,
+                             '-c', 'copy', output_path]
+        subprocess.run(ffmpeg_concat_cmd)
+
+    return output_path
+
+
+def encode_video(video: np.ndarray, output_path: str,
+                 fps: float, bitrate: str = "0", crf: int = 20) -> str:
+    """Encode a video with vp9 codec via imageio-ffmpeg
+
+    Parameters
+    ----------
+    video : np.ndarray
+        Video to be encoded
+    output_path : str
+        Desired output path for encoded video
+    fps : float
+        Desired frame rate for encoded video
+    pix_fmt_out : str
+        Desired pixel format for output encoded video, by default "yuv420p"
+    bitrate : str, optional
+        Desired bitrate of output, by default "0". The default *MUST*
+        be zero in order to encode in constant quality mode. Other values
+        will result in constrained quality mode.
+    crf : int, optional
+        Desired perceptual quality of output, by default 20. Value can
+        be from 0 - 63. Lower values mean better quality (but bigger video
+        sizes).
+
+    Returns
+    -------
+    str
+        Output path of the encoded video
     """
 
-    # ffmpeg expects the video shape in width, height not row, col
-    # have to reverse shape when inputting
-    # gray8 is uint8 format
+    # ffmpeg expects video shape in terms of: (width, height)
+    video_shape = (video[0].shape[1], video[0].shape[0])
 
     writer = mpg.write_frames(output_path,
-                              (video[0].shape[1], video[0].shape[0]),
+                              video_shape,
                               pix_fmt_in="gray8",
                               pix_fmt_out="yuv420p",
                               codec="libvpx-vp9",
                               fps=fps,
-                              bitrate=bitrate)
-    writer.send(None)
+                              bitrate=bitrate,
+                              output_params=["-crf", str(crf)])
+
+    writer.send(None)  # Seed ffmpeg-imageio writer generator
     for frame in video:
         writer.send(frame)
     writer.close()
+
+    return output_path
+
+
+def transform_to_webm(video: np.ndarray, output_path: str,
+                      fps: float, ncpu: int, bitrate: str = "0",
+                      crf: int = 20) -> str:
+    """Function to transform 2p gray scale video into a webm
+    video using imageio_ffmpeg.
+
+    Parameters
+    ----------
+    video : np.ndarray
+        Video to be transformed with shape (time, row, col)
+    output_path : str
+        Output path for the transformed video
+    fps : float
+        Desired frames per second (fps) of the output video
+    ncpu : int
```

Comment:
I think there should be a mechanism for setting this to [the number of cpus on the system](https://docs.python.org/3.4/library/multiprocessing.html?highlight=process#multiprocessing.cpu_count). That way, if we are running jobs with `#PBS -n` it will utilize all the CPUs on a node whether it lands on a 24, 32, or 40.
I think a value of `-1` could suffice to trigger that behavior.

Reply:
Unclear whether the number of python `multiprocessing` module is smart enough to realize that not all CPUs are available. I think it's safer to require specifying `ncpu` in the actual qsub script where you can specify `#PBS -l ncpus=#`.

Reply:
@djkapner Yeah this was by my request, since if we run on the cluster not all on the node are necessarily available.